semantic-core/src/Analysis/Concrete.hs

```diff
@@ -74,46 +74,39 @@ concrete
 runFile :: ( Carrier sig m
            , Effect sig
            , Member Fresh sig
-           , Member (Reader FrameId) sig
            , Member (State Heap) sig
            )
         => File (Term Core.Core User)
         -> m (File (Either (Loc, String) Concrete))
 runFile file = traverse run file
   where run = runReader (fileLoc file)
             . runFailWithLoc
+            . runReader (mempty :: Env)
             . fix (eval concreteAnalysis)
 
 concreteAnalysis :: ( Carrier sig m
                     , Member Fresh sig
+                    , Member (Reader Env) sig
```

Comment:
Rather than modelling the local environment explicitly in the heap, we resume using an `Env` for it, and modelling non-local scopes explicitly with records. Computation of scope graphs is still possible by instrumenting `bind`s (among other things), but we get to enjoy a much simpler implementation of name binding and lookup.